src/fastfilter.h

```diff
@@ -7,10 +7,121 @@
+#define LN2 0.6931471805599453094172321214581765680755001343602552
+#define MIN_N_HASH_FUNC 1
+#define MAX_N_HASH_FUNC 32
 class uint256;
 
+/**
+ * This class can be used anywhere a Bloom filter is used so long as the input data is random.
+ *
+ * If nHashFuncs is 16 and nFilterItems is >= 64k all bits in the uint256 input data will be used to set bits in
+ * the filter.
+ *
+ * nHashFuncs may range from 2 to 32 inclusive.
+ */
+class CVariableFastFilter
+{
+protected:
+    // A bit vector containing the bloom filter data
+    std::vector<unsigned char> vData;
+
+public:
+    uint8_t nHashFuncs;
+    uint32_t nFilterBytes;
+    uint64_t nFilterItems;
+
+    CVariableFastFilter() : nHashFuncs(2), nFilterItems(2){};
+
+    CVariableFastFilter(uint64_t nElements, double nFPRate)
+    {
+        if (nElements == 0)
+        {
+            LOGA("Construction of empty CVariableFastFilter attempted.\n");
+            nElements = 1;
+        }
+
+        nFilterBytes = (uint32_t)(std::ceil(-1 / LN2SQUARED * nElements * log(nFPRate) / 8));
+        nFilterItems = 8 * nFilterBytes;
+
+        if (nFilterBytes > std::numeric_limits<uint32_t>::max())
+            throw std::runtime_error("CVariableFastFilter can have size no greater maximum uint32_t.");
+
+        vData.resize(nFilterBytes, 0);
+
+        unsigned int optimalNHashFuncs =
+            (unsigned int)std::max(MIN_N_HASH_FUNC, int(nFilterBytes * 8 / nElements * LN2));
+        if (optimalNHashFuncs > MAX_N_HASH_FUNC)
+        {
+            LOGA("CVariableFastFilter constructed with suboptimal number of hash functions.\n");
+            nHashFuncs = MAX_N_HASH_FUNC;
+        }
+        else
+            nHashFuncs = optimalNHashFuncs;
+    }
+
+
+    // returns true IF this function made a change (i.e. the value was previously not set).
+    bool checkAndSet(const uint256 &hash)
+    {
+        const uint32_t *pos = (const uint32_t *)hash.begin();
+        bool unset = 0; // If any position is not set, then this will be true
+        for (unsigned int i = 0; i < nHashFuncs; i++, pos++)
+        {
+            uint32_t val = *pos;
+            uint32_t idx = val % (nFilterItems - 1);
+            uint32_t bit = (1 << (idx & 7));
+            idx >>= 3;
+            unset |= (0 == (vData[idx] & bit));
+            vData[idx] |= bit;
+        }
+
+        return unset;
+    }
+
+    void insert(const uint256 &hash)
+    {
+        const uint32_t *pos = (const uint32_t *)hash.begin();
+        for (unsigned int i = 0; i < nHashFuncs; i++, pos++)
+        {
+            uint32_t val = *pos;
+            uint32_t idx = val % (nFilterItems - 1);
+            vData[idx >> 3] |= (1 << (idx & 7));
+        }
+    }
+
+    bool contains(const uint256 &hash) const
+    {
+        const uint32_t *pos = (const uint32_t *)hash.begin();
+        bool unset = 0; // If any position is not set, then this will be true
+        for (unsigned int i = 0; i < nHashFuncs; i++, pos++)
```

Comment:
when we do pos++ we are adding 4 bytes because pos is a pointer to a uint32_t.  With MAX_N_HASH_FUNC == 32, we'll overrun the hash which only has 32 bytes, not 4*32 bytes.  

Is MAX_N_HASH_FUNC == 8 acceptable for very large blocks?  If not, then every 8 iteratons of the loop we need to permute hash.  A relatively fast permutation would be to rotate the whole thing by 8 bits via: char tmp = hash.begin(); memcpy(...); hash.begin()[31] = tmp;

Reply:
Actually, it's the small IBLTs that require larger numbers of hash functions for optimality. Unfortunately, in iblt_params.h an IBLT accommodating 7 entries requires 9 hash functions. The rest require 8 or fewer. One simple fix would be to just nudge that value down to 8 and then enforce MAX_N_HASH_FUNC == 8. It's very unlikely that this small change will greatly affect the optimal size of the IBLT.

On the other hand, iblt_params.h is autogenerated and tuned to a specific decode rate. Perhaps one day we will want to adjust that, or someone will innocently regenerate it without knowing about the tweak. So I could also see the argument to just implement the permutation as you suggested. I see the pattern you are suggesting, so I'll play around to see if I can get it compiled.

Reply:
Ok, I took a shot at what you recommended regarding hash rotation. Tests are passing and it seems to be working correctly on mainnet, but please take an extra close look at what I did :-).

Reply:
Hm, strange. Looks like fast filter tests are failing on Travis but not on my machine. I'll see if I can reproduce it.